xarray_sentinel/sentinel1.py

```diff
@@ -141,34 +142,115 @@ def open_orbit_dataset(product_path: PathType) -> xr.Dataset:
-    product_attrs["groups"] = []
-    for sub_swath in sub_swaths:
-        for data in METADATA_OPENERS:
-            product_attrs["groups"].append(f"{sub_swath}/{data}")
+def find_avalable_groups(
+    manifest_path: PathType,
+    product_attrs: T.Dict[str, T.Any],
+    product_files: T.Dict[str, str],
+) -> T.Dict[str, T.Dict[str, T.Any]]:
+    ancillary_data_paths = esa_safe.get_ancillary_data_paths(
+        manifest_path, product_files
+    )
+    groups: T.Dict[str, T.Dict[str, T.Any]] = {}
+    for subswath_id, subswath_data in ancillary_data_paths.items():
+        subswath_id = subswath_id.upper()
+        burst_info = get_burst_info(product_attrs, subswath_id, subswath_data)
+        subgroups = list(METADATA_OPENERS.keys()) + list(burst_info.keys())
+        groups[subswath_id] = dict(subswath_data, subgroups=subgroups)
+        for subgroup in METADATA_OPENERS.keys():
+            groups[f"{subswath_id}/{subgroup}"] = subswath_data
+        for burst_id, burst_data in burst_info.items():
+            full_data = dict(subswath_data, **burst_data)
+            groups[f"{subswath_id}/{burst_id}"] = full_data
+    return groups
+
+
+def open_root_dataset(
+    product_attrs: T.Dict[str, str], groups: T.Dict[str, T.Dict[str, T.Collection[str]]]
+) -> xr.Dataset:
+    product_attrs = dict(product_attrs, groups=list(groups.keys()))
     return xr.Dataset(attrs=product_attrs)  # type: ignore
 
 
-def open_swath_dataset(product_path: str, group: str) -> xr.Dataset:
-    product_attrs = {"groups": ["orbit", "attitude", "gcp"]}
+def open_swath_dataset(
+    product_attrs: T.Dict[str, str],
+    swath_id: str,
+    swath_data: T.Dict[str, T.Collection[str]],
+) -> xr.Dataset:
+    product_attrs = dict(product_attrs, groups=swath_data["subgroups"])
     return xr.Dataset(attrs=product_attrs)  # type: ignore
 
 
+def open_burst_dataset(
+    product_attrs: T.Dict[str, str], burst_id: str, burst_data: T.Dict[str, T.Any]
+) -> xr.Dataset:
+    ds = xr.merge(
+        [
+            {
+                pol.upper(): rioxarray.open_rasterio(datafile)
+                for pol, datafile in burst_data["measurement"].items()
+            }
+        ]
+    )
+    ds.attrs.update(product_attrs)
+    ds = ds.squeeze("band").drop(["band", "spatial_ref"])
+    return ds.isel(
+        x=slice(burst_data["burst_first_pixel"], burst_data["burst_last_pixel"] + 1),
+        y=slice(burst_data["burst_first_line"], burst_data["burst_last_line"] + 1),
+    )
+
+
+def get_burst_id(
+    product_attrs: T.Dict[str, T.Any], burst_centre: np.typing.ArrayLike,
+) -> str:
+    rounded_centre = np.int_(np.round(burst_centre, 1) * 10)
+    n_or_s = "N" if rounded_centre[0] >= 0 else "S"
+    e_or_w = "E" if rounded_centre[1] >= 0 else "W"
+    return (
+        f"R{product_attrs['sat:relative_orbit']:03}"
+        f"-{n_or_s}{rounded_centre[0]:03}"
+        f"-{e_or_w}{rounded_centre[1]:04}"
+    )
+
+
+def get_burst_info(
```

Comment:
Same as above. `get_` is reserved for function that actually *get* the information. The logic to get the information may be complex, but the information is not manipulated.

Reply:
In this case I find `make_burst_info` and `build_burst_info` a bit clumsy: you don't make information (hopefully!), you just read and format it. Considering that I was planning to use `open_gcp_dataset` to load the GCPs instead of duplicating the code, what about `gather_`? It gives the idea that you collect data from many places.